src/rover/camera.cpp

```diff
@@ -0,0 +1,229 @@
+        session->buffers[i].data = new uint8_t[session->image_size];
+    }
+
+    // Link each buffer with the driver.
+    for (size_t i = 0; i < NUM_BUFFERS; i++) {
+        // Allocate a V4L2 "buffer".
+        // This really isn't a buffer, but a struct that contains information
+        // about the buffers we will provide.
+        struct v4l2_buffer vbuf;
+        memset(&vbuf, 0, sizeof(vbuf));
+
+        vbuf.type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
+        // USERPTR mode: provide our own pointers.
+        vbuf.memory = V4L2_MEMORY_USERPTR;
+        // Which buffer is this?
+        vbuf.index = i;
+        // Pointer to buffer's data.
+        vbuf.m.userptr = (unsigned long) session->buffers[i].data;
+        // How big is the buffer?
+        vbuf.length = session->buffers[i].size;
+
+        // Link the buffer.
+        if (ioctl(session->fd, VIDIOC_QBUF, &vbuf) != 0) {
+            return Error::LINK_BUFFERS;
+        }
+    }
+
+    // Initialize the frame buffer.
+    session->frame_buffer = new uint8_t[session->image_size];
+
+    return Error::OK;
+}
+
+Error start(CaptureSession* session) {
+    // Set an enum value to let V4L2 know that we want to start video capture.
+    enum v4l2_buf_type buf_type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
+
+    // Start the video stream.
+    if (ioctl(session->fd, VIDIOC_STREAMON, &buf_type) != 0) {
+        return Error::START_STREAM;
+    }
+
+    return Error::OK;
+}
+
+Error grab_frame(CaptureSession* session, uint8_t** out_frame, size_t* out_frame_size) {
+    // Set up structs for the select call.
+    fd_set fds;
+    struct timeval tv;
+
+    // Set up our file descriptors to watch (only fd).
+    FD_ZERO(&fds);
+    FD_SET(session->fd, &fds);
+
+    // Set the timeout.
+    tv.tv_sec = SELECT_TIMEOUT;
+    tv.tv_usec = 0;
+
+    // Wait for the next frame. See man select(2) for more information.
+    if (select(session->fd + 1, &fds, NULL, NULL, &tv) <= 0) {
+        return Error::SELECT;
+    }
+
+    // Set up the video4linux "buffer" (which points to our buffer).
+    struct v4l2_buffer vbuf;
+    memset(&vbuf, 0, sizeof(vbuf));
+    
+    vbuf.type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
+    vbuf.memory = V4L2_MEMORY_USERPTR;
+
+    // Read the frame.
+    if (ioctl(session->fd, VIDIOC_DQBUF, &vbuf) != 0) {
+        return Error::READ_FRAME;
+    }
+
+    // Get the pointer to our frame data.
+    *out_frame = (uint8_t*) vbuf.m.userptr;
+
+    // Get the actual size of the frame data.
+    *out_frame_size = vbuf.bytesused;
+
+    // Set the buffer we used!
+    session->last_capture_buffer = vbuf;
+
+    return Error::OK;
+}
+
+Error return_buffer(CaptureSession* session, uint8_t* buffer) {
+    // Reset the buffer we just used.
+    if (ioctl(session->fd, VIDIOC_QBUF, &session->last_capture_buffer) != 0) {
+        return Error::PREPARE_BUFFER;
+    }
+
+    return Error::OK;
+}
+
+void close(CaptureSession* session) {
+    ::close(session->fd);
```

Comment:
This looks a bit weird... not sure what it does

Reply:
close is a Linux syscall which lives in global scope. The "::" tells C++ to look at that one (so that it doesn't call close recursively)